### ISSUE TITLE
balena tunnel: [Error] connect ETIMEDOUT

### ISSUE THREAD

**Issue by pdcastro** [balena-io/balena-cli#2172]
The `balena tunnel` command in CLI releases older than v12.38.5 result in ETIMEDOUT:
```
$ balena tunnel 6b1e674 -p 22222
[Info]    Opening a tunnel to 6b1e674339ff7df836ac92c686e938ba...
[Info]     - tunnelling localhost:22222 to 6b1e674339ff7df836ac92c686e938ba:22222
[Info]    Waiting for connections...
[Error]   127.0.0.1:58599 => 127.0.0.1:22222 ===> 10.240.0.30:22222 :: connect ETIMEDOUT 3.224.154.111:3128
```

**The solution / fix is to update the CLI to v12.38.5 or later.** The newer CLI releases take advantage of a new backend endpoint with TLS encryption (`tunnel.balena-coud.com:443`). Older releases attempt to use an endpoint that is no longer available for security reasons (`vpn.balena-cloud.com:3128`), hence the timeout error.

**Comment by pdcastro** [comment]
In the case of openBalena, ensure that CLI v12.38.5 or later is used with openBalena v3.1.2 or later, and use older CLI releases with older openBalena releases. Instructions on updating openBalena have recently been added to the [openBalena Getting Started guide](https://www.balena.io/open/docs/getting-started/).

**Comment by mathvav** [comment]
@pdcastro Are there any plans to better alert Balena customers to update their CLI tools for future issues like this?

I was on a CLI version from early December and the tool failed to operate--I spent about two hours on this issue before stumbling on this issue on my way to file a bug report after talking to my network admin and moving the device to a different building (different network). **I'm using Balena Cloud.**

I'm not sure this change was planned and am not sure how atypical this type of change is, but if it's a thing where 2 month old releases can break functionality and look like network errors, I would hope that Balena would at least have some sort of message in the command line tool that an update is available, and make updates easier than going to the GitHub to download a new installer.

Again: I know stuff happens, and maybe this is an issue that only impacts a single-digit number of users, or maybe nothing like this has happened before, but it feels like something like an CLI update available message (similar to NPM or pip or many other tools) or some tool on the Balena dashboard is a reasonable ask when this types of thing happens.

**Comment by pdcastro** [comment]
@mathvav, I apologize for your experience as a result of this issue with the `balena tunnel` command.

This issue was indeed atypical. We considered modifying the balenaCloud backend to return a user-friendly error message to old CLI clients at the HTTP level, so that the `balena tunnel` command would print something along the lines of _"please update the CLI."_ However, because of how the old CLI clients were coded, by the time that the friendly error message was printed, the CLI had already sent credentials in clear text to the backend (security issue #2042). We found that the only way of ensuring that the old clients would not send clear text credentials was to prevent the connection from being established at the TCP level, which then led to the ETIMEDOUT error. Unfortunately, the security patch came at the cost of user experience.

I think that something that we failed to fully grasp was that the ETIMEDOUT error would lead users to assume that the problem was in their local network, rather than a problem with the CLI or balenaCloud. It might have tilted the balance towards continuing to allow clear-text credential requests, at least for a short period of time, with a friendly error message. This is a lesson we have learned for the future. Thank you for sharing your experience. :+1:

Thank you also for the improvement suggestions and feedback regarding notification mechanisms (CLI and web dashboard) and the CLI installation / update process. We have started a discussion regarding a notification mechanism through the CLI console (issue #2178) so that, in the future, old CLI clients could warn users about security issues. The CLI already has an "update available" notification mechanism ([update.ts](https://github.com/balena-io/balena-cli/blob/master/lib/utils/update.ts)), and I will double check that it is doing its job across all installation options and supported platforms. We also have plans of making updates easier with a `balena update` command that automatically downloads and executes installers. On Linux, we also plan to provide an `apt install balena-cli` workflow. Fingers crossed that we will be able to release these improvements soon. Thanks for working with us along the way!

**Comment by mathvav** [comment]
Thanks for the update! It makes more sense hearing the cause of the issue being security--that is understandable. And, the `balena update` looks like something to look forward to!

However, as per my original message, the update messages do not appear to be working for me. I see the `notify.ts` file, but it doesn't seem to be working on my Windows work computer (currently running 12.38.8, about a week old, and I've _never_ seen a banner). I also checked on my personal laptop (macOS) as I had played around with Balena for a side project (and it's running `12.37.0`--no message).

My Windows computer I've tried both Git Bash, command prompt, and PowerShell--my Mac is running zsh in iTerm 2. I'm not sure what is happening, but I'd be happy to provide debugging information if necessary.

**Comment by mathvav** [comment]
This makes it look like it checked 30 minutes ago (Windows computer, Git Bash shell):

```shell
$ cat ~/.config/configstore/update-notifier-balena-cli.json
{
        "optOut": false,
        "lastUpdateCheck": 1612987074356
}
```

**Comment by mathvav** [comment]
I'm happy to dig a little further since it is open source, but can you first confirm that a banner is displayed correctly on your side?

**Comment by pdcastro** [comment]
Hey @mathvav, it's great to hear that you would like to help with debugging the notification mechanism. Yes, I can confirm that I tested it on all platforms (Linux, Mac, Windows), all installers (GUI installer, standalone zip package, npm installation), and a few shells (on Windows: PowerShell, command prompt, MSYS2). It worked as expected in all cases. Your experience makes me wonder that my tests were not representative of real usage for some reason, so yes I'd like to get this debugged!

> it seems as it's designed to check once a day, and cache that result for subsequent runs of the command

That's right, and you have already found the file where the last check timestamp is stored: `~/.config/configstore/update-notifier-balena-cli.json`. When more than 24 hours have passed since the last check, it will check again.

***One caveat is that the check/test is skipped if the user is `root` on Linux or macOS*** (presumably to avoid permission errors when $HOME points to a non-root homedir and the last check timestamp needs updating), but this should not apply to Windows. ***(WSL is considered to be Linux for the purposes of that test.)***

Here's how I had tested it the other day:
1. Open `~/.config/configstore/update-notifier-balena-cli.json` in a text editor.
2. Replace the large value of `lastUpdateCheck` with 1, or subtract 86,400,000 (milliseconds in a day) from it, so that the file looks like:
```
{
	"optOut": false,
	"lastUpdateCheck": 1
}
```
3. Run some CLI command like `balena version`.
4. _If the current CLI version is older than [the latest version available in the npm registry](https://www.npmjs.com/package/balena-cli),_ then the contents of `~/.config/configstore/update-notifier-balena-cli.json` will be changed to something like:
```
{
	"optOut": false,
	"lastUpdateCheck": 1613002032495,
	"update": {
		"latest": "12.40.0",
		"current": "12.39.1",
		"type": "minor",
		"name": "balena-cli"
	}
}
```
5. Run the CLI again, e.g. `balena version`. Again, _if the current CLI version is older than [the latest version available in the npm registry](https://www.npmjs.com/package/balena-cli),_ then the CLI will print this banner:
```
$ balena version

   ╭────────────────────────────────────────────────────────────────────╮
   │                                                                    │
   │                 Update available 12.39.1 → 12.40.0                 │
   │                                                                    │
   │   https://github.com/balena-io/balena-cli/blob/master/INSTALL.md   │
   │                                                                    │
   ╰────────────────────────────────────────────────────────────────────╯

12.39.1
```
and the contents of `~/.config/configstore/update-notifier-balena-cli.json` will be changed to something like the following (the timestamp doesn't change, but the "update" property disappears):
```
{
	"optOut": false,
	"lastUpdateCheck": 1613002032495
}
```

If the current version of the CLI is already the same as the latest version in the npm registry, then the "update" property is not added to `~/.config/configstore/update-notifier-balena-cli.json`, but the timestamp is still updated once every 24 hours.

Presumably some of the steps above don't match what you observe. The latest version of the CLI on the npm registry is currently 12.40.0. What's your CLI version, and what step above does not match? Were you running it as the root user on WSL?

**Comment by pdcastro** [comment]
Closing this issue in relation to the ETIMEDOUT error that was resolved in CLI version 12.38.5 or later.
With regard to the missing update banner, if it is still happening following my testing procedure above, I suggest a new specific issue to be opened. Thanks for the feedback!
